lib/Extractor/KLEEBuilder.cpp

```diff
@@ -169,6 +172,21 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
                                                                                   klee::ConstantExpr::create(i, Width))));
       PowerTwoBitsMap[Origin] = PowerExpr;
     }
+    if (Origin->NumSignBits > 1) {
+      ref<Expr> SignBitsExpr = klee::ConstantExpr::create(1, Expr::Bool);
+      ref<Expr> Res = LShrExpr::create(Var, klee::ConstantExpr::create(Width-Origin->NumSignBits, Width));
+      ref<Expr> TestOnes = SubExpr::create(ShlExpr::create(klee::ConstantExpr::create(1, Width),
+                                           klee::ConstantExpr::create(Origin->NumSignBits, Width)),
+                                           klee::ConstantExpr::create(1, Width));
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     OrExpr::create(EqExpr::create(Res, TestOnes),
+                                                    EqExpr::create(Res, klee::ConstantExpr::create(0, Width))));
+      ref<Expr> MSB = ExtractExpr::create(Var, Width-1, Expr::Bool);
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     Expr::createImplies(EqExpr::create(MSB, klee::ConstantExpr::create(0, Expr::Bool)),
+                                                         EqExpr::create(ctlz(Var), klee::ConstantExpr::create(Origin->NumSignBits, Width))));
```

Comment:
I compared for "N" higher bits only and thus logically shifted them. Arithmetic shift would lead to comparison with all bits. I have updated it to arithmetic shift though.
The second condition to build sign bits expression is to make sure that the leading number of zeros should be equal to number of sign bits for positive variables.
If I don't teach Souper about this condition, then we can't validate test cases #6, #8, #10, #11. I just tried to make souper more stronger herethan LLVM.